### DISCUSSION TITLE
`killed signal terminated program cc1plus` building OAK-D Lite packages on RPi4

### DISCUSSION THREAD

**Discussion by jrhanobik** [iRobotEducation/create3_docs#135] [Questions] [answered]
Has anyone had any luck installing an OAK-D Lite on a Raspberry Pi 4 (4GB Memory)?  I have tried several attempts following this document -> https://robofoundry.medium.com/oak-d-lite-camera-ros2-setup-1e74ed03350d but in all cases, the build fails while building the depthai_bridge.  It's gets to 66% then just hangs for awhile and finally dies.

**Comment by shamlian** [comment]
While I've never tried to build this particular package, I have experienced `colcon` starving itself when trying to build large things on Raspberry Pi. To get around this, reduce the number of cores for `make` and tell `colcon` to process jobs in sequence and not in parallel. Try `MAKEFLAGS="-j3" colcon build --executor sequential` instead of just `colon build`.

For further information, see the colcon docs [here](https://colcon.readthedocs.io/en/released/reference/executor-arguments.html).

**Reply by jrhanobik** [reply]
I gave that a shot and it still failed.  This time, depthai_bridge got to 50% and died.  see below.

![image](https://user-images.githubusercontent.com/4820984/183453180-19b92f29-fb6a-473f-87fc-f7d22dc35cae.png)

**Reply by alsora** [reply]
As Steve mentioned above, you are running out of RAM while building.
Make sure that you are not running any additional application on the Raspberry Pi while compiling.
I suggest you to reduce the number of concurrent jobs even more, from 3 to only 1.

I know of several packages that can exhaust all RAM on a RaspberryPi even when running with only 1 concurrent job.
If that's the case, I recommend you to report the issue to the maintainers of this package.

In the meantime, there are a variety of methodologies to address these kind of problems, but they require a very advanced knowledge of C++ compilation.
You could investigate different linkers (e.g. `gold` rather than `ld`) and maybe even different build programs (e.g. `ninja` rather than `make`).
Lastly you have access to a wide variety of linking flags that can be specified via CMake and some of them will reduce the amount of RAM required.

Unfortunately, I really can't give you precise advices on how to implement all of this.

You can also look at the content of this package.
If there are examples or other parts that you don't need, comment them out to save resources.


Another option is to do cross-compilation.
This way you can build all your packages on your laptop and just copy them to the Raspberry Pi.
Also cross-compilation is a relatively complex topic, however as a relatively quick solution, you could use an arm64 docker container running the same OS as your Raspberry Pi.
Executables and libraries built in this docker will be compatible with your raspberry Pi.
This is not a real cross-compilation, but it will get the job done.

**Reply by justinIRBT** [reply]
The highest level of restriction with the build line is:
`MAKEFLAGS="-j1 -l1" colcon build --executor sequential `
That will limit both building and linking to 1 core

**Reply by jerrynsr** [reply]
I built this successfully on a RPI4 w/8GB of RAM.   You can configure your OS install to boot from USB3 SSD and allocate 4GB of Swap. 

Since I build ROS2 packages on the target and use Colcon I made sure my 8GB RPI4 isn't starved for memory by adding an 8GB of SWAP.

**Reply by jrhanobik** [reply]
Although I've been able to successfully build the demos, launching the example (ros2 launch depthai_examples stereo.launch.py) it starts up just fine, runs a few seconds and then dies with the following error:

[INFO] [robot_state_publisher-2]: process has finished cleanly [pid 2722]
[INFO] [component_container-4]: process has finished cleanly [pid 2726]
[INFO] [component_container-3]: process has finished cleanly [pid 2724]
[stereo_node-1] [18443010512FAF1200] [1.1.1] [111.464] [MonoCamera(1)] [error] OV7251 only supports THE_480_P/THE_400_P resolutions, defaulting to THE_480_P
[stereo_node-1] [18443010512FAF1200] [1.1.1] [111.464] [MonoCamera(0)] [error] OV7251 only supports THE_480_P/THE_400_P resolutions, defaulting to THE_480_P
[stereo_node-1] terminate called after throwing an instance of 'rclcpp::exceptions::RCLError'
[stereo_node-1]   what():  could not create service: rcl node's context is invalid, at /tmp/binarydeb/ros-galactic-rcl-3.1.3/src/rcl/node.c:426
[ERROR] [stereo_node-1]: process has died [pid 2720, exit code -6, cmd '/home/ubuntu/luxonis_depthai_ws/install/depthai_examples/lib/depthai_examples/stereo_node --ros-args --params-file /tmp/launch_params_82om509o --params-file /tmp/launch_params_ohme8049 --params-file /tmp/launch_params_9nlblu1l --params-file /tmp/launch_params_cyhoat3c --params-file /tmp/launch_params_cckx7f34 --params-file /tmp/launch_params_c81uwead --params-file /tmp/launch_params_f07nb5gm --params-file /tmp/launch_params_o8noz4ym'].


Have you been able to run similar?

**Reply by jerrynsr** [reply]
I am almost to the point of testing this.  I built the demos,  integrated the ClearPath Robotics TurtleBot4 code .. I have NAV2 working great.  I just need to get all the parts together to power up the Oak D Lite camera.     How are you powering up the camera ?  I did tried powering up with my RPI4 but the USB Audio, SlamTec LIDAR, and Addata SSD all push the PI4 USB delivery to the limit.

**Reply by jrhanobik** [reply]
Right now, I've been trying to introduce sensor by sensor.  For this exercise, I created a whole new SD card with a fresh ubuntu / ros2 install and then installed just the OAK-D Lite.  For this try, was just powering off of the RP4.  considering a powered USB Hub don't want to understand the limits of the onboard USB as I add more.  I also have an RPLidar that not's hooked at the moment but that's it for now.  When you said integrated the Clearpath code, specifically from source OR did you burn one of the ready made images?

**Reply by jerrynsr** [reply]
My Create3 development environment resides completely in the Create3 robot.  I have an 8GB Raspberry PI 4 installed that boots and runs off an Addata M.2 SSD.  I have a small profile aluminum m.2 to USB 3.0 enclosure in the bot.  All my ROS workspace source code  was checked out from GitHub.  All code was built on the RPI4.   My current georgebot_ws  includes paclages from iRobot and ClearPath Robotics.  It's still a work in progress but I have mapping and navigation using saved maps of my house working.   Compile times on the Raspberry PI4 using an SSD drive isn't too bad.  It's much faster than compiling on an SD Card.

I am not using ready made images.   I'm using the Ubuntu 20.04 IoT image.  I modified it so it does not connect through the Ubuntu Iot cloud..    I have it configured to boot and grab an ip address from my network.  I followed the Ubuntu 20.04 ROS2 Galactic install.
My current georgebot_ws consists of source code from both iRobot Create3 Repos and  ClearPath Robotics.

**Comment by jrhanobik** [comment, marked as answer]
I appreciate the several posts that pointed out that this was caused by memory starvation and that made me reevaluate my approach.  I was following a doc I found that suggested to start with a fresh install of ubuntu / ros2 etc. so I did.  The one place I fell short in my desire to get things rolling was that I neglected to create a swap file that once created on a new install, worked flawlessly!  Thanks for the input that pointed me in the right direction.
